File: treequesiton/LevelOrder.java
import java.util.*;

public class LevelOrder {

    // Definition for a binary tree node
    static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int val) {
            this.val = val;
        }
    }

    // Function to calculate average of each level
    public static List<Double> averageOfLevels(TreeNode root) {

        List<Double> ans = new ArrayList<>();

        // Edge case
        if (root == null) return ans;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        // BFS traversal
        while (!queue.isEmpty()) {

            int levelSize = queue.size(); // number of nodes at this level
            double sum = 0;

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                sum += node.val;

                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }

            // Average of current level
            ans.add(sum / levelSize);
        }

        return ans;
    }

    // Optional: main method for testing
    public static void main(String[] args) {

        /*
            Tree:
                 3
                / \
               9  20
                  / \
                 15  7
        */

        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println(averageOfLevels(root));
    }
}

File: treequesiton/levelOrder.js
function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val);
    this.left = (left === undefined ? null : left);
    this.right = (right === undefined ? null : right);
}

const root = new TreeNode(10);
root.left = new TreeNode(6);
root.right = new TreeNode(15);

root.left.left = new TreeNode(3);
root.left.right = new TreeNode(8);

root.right.right = new TreeNode(20);

var levelOrderSuccessor = function(root, key) {

    if (root === null) return null;

    let queue = [];
    queue.push(root);

    while (queue.length > 0) {

        let node = queue.shift();

        if (node.left !== null) queue.push(node.left);
        if (node.right !== null) queue.push(node.right);

        if (node.val === key) {
            return queue.length > 0 ? queue[0].val : null;
        }
    }

    return null;
};

console.log(levelOrderSuccessor(root, 3)); // 8
